ColorInspectorTool : Move out of ImageView into tool

danieldresser-ie opened the pull request:
This is an initial pass on making the Color Inspector into a tool that can be toggled off and on, instead of a built-in part of ImageView.

So far, this has just been about shuffling the code around and making sure everything stills works - I haven't reviewed everything for whether there are now better ways of doing things ... I suspect there are some things that could be clarified now if I go through it with a fine comb.

Known issues currently:
* needs an icon ( I'm happy to take a swing at making an icon, but I figured I should see if you have an idea for what it should be first )
* needs a hotkey
* There is a mess involved with finding the right image plug to evaluate. We want an image that has all the preprocessing built into ImageView done ( picking the right view, flattening the image ), but we don't want to preprocessing added by users to affect the color values you're inspecting ( since we assume that if users are adding something to the preprocessing stack, it's probably functioning as some sort of view transform ?? ). In the existing code, there is a bit of a hack to do this: we use `view->getPreprocessor()->getChild<DeepState>( "__flattenedImage" )->getChild<ImagePlug>( "out" )` to find the preprocessing node with a specific name. This is now a problem because now this code has now been moved outside of ImageView, so we no longer have access to the protected `getPreprocessor` ( I've crudely hacked it to be public for the moment, but that's clearly not an actual solution ). It feels like there ought to be a cleaner way to define what we actually want here. There's also some kinda ugly looking code in Python that is trying to accomplish something similar: _ColorInspectorPlugValueWidget does `self.getPlug().node().view().viewportGadget().getPrimaryChild().getImage()` in order to find an image it can evaluate to find a list which channels exist ( here, it doesn't matter whether the colors have been processed, but we do want the correct view to have been selected already ).

johnhaddon commented on python/GafferImageUI/ColorInspectorToolUI.py:
```diff
@@ -0,0 +1,673 @@
+
+	"description",
+	"""
+	The default tool for an image view - shows the color values for what is under the cursor, and allows
+	Ctrl-clicking or Ctrl-dragging to create persistent pixel or region inspectors.
```
Oh, and we should include the no-modifier drag to drag pixel values here too.

danieldresser-ie commented on python/GafferSceneUI/ImageSelectionToolUI.py:
```diff
@@ -46,6 +46,18 @@
+	Tool for selecting objects based on image data. Requires an `id` image layer
+	from a Gaffer render, with gaffer:renderManifestFilePath metadata, or an
+	ObjectID cryptomatte, or an `instanceID` image layer for selecting numerical
+	ids of instances within an encapsulated instancer.
```
I'll defer this commit to the next PR where the instanceID feature does exist.

danieldresser-ie commented on python/GafferImageUI/ColorInspectorToolUI.py:
```diff
@@ -0,0 +1,673 @@
+			"description",
+			"""
+			Display the value of the pixel under the cursor. Ctrl-click to add an inspector to a pixel, or
+			Ctrl-drag to create an area inspector. Display shows value of each channel, hue/saturation/value, and Exposure Value which is measured in stops relative to 18% grey.
```
I've made mode specific tooltips. I've removed the mention of Ctrl-Drag completely ... it's not specific to this widget, so having it documented just on the main tooltip is probably fine?

danieldresser-ie commented on src/GafferImageUI/ColorInspectorTool.cpp:
```diff
@@ -0,0 +1,1067 @@
+	// because derived classes may have called insertConverter(),
+	// so we take it from the input to the display transform chain.
+
+	ImagePlug *image = view->getPreprocessor()->getChild<DeepState>( "__flattenedImage" )->getChild<ImagePlug>( "out" );
```
I've switched to using ImageGadget.getImage() in both places. Only problem is that getImage is const, and setInput requires a non-const plug ... do you thinking cheating that with a const-cast is illegitimate?

johnhaddon replied:
I'm not too concerned about that, but it looks like we could add a non-const accessor to ImageGadget if you wanted.

danieldresser-ie commented:
I think I've now addressed all feedback, and had a quick go at a possible icon ... but I'm still not feeling 100% today, so I might have missed something, I'll double check things on Monday.

johnhaddon commented:
> taken another swing at the icon ( not in love with it...

It feels like an improvement to me. I still felt like the red was a part of the circle overlay rather than a pixel underneath it. And the pixels feel a bit like a regular bayer pattern than something that a plausible chunk of an image. I tried desaturating them and making them all a bit more similar to see if that helped : https://github.com/johnhaddon/gaffer/commit/4d7b5067e3dc272331f7f73d32e9fee72569c314. Also not in love, but might help progress the conversation at least.

danieldresser-ie commented on python/GafferImageUI/ColorInspectorToolUI.py:
```diff
@@ -0,0 +1,693 @@
+				GafferImageUI.ColorInspectorTool.ColorInspectorPlug.Mode.Pixel : "Displays the color values of pixel inspector %s." % plug.getName()[9:],
+				GafferImageUI.ColorInspectorTool.ColorInspectorPlug.Mode.Area : "Displays the average color values of region inspector %s." % plug.getName()[9:],
+			}[ plug["mode"].getValue() ] + """
+			Display shows value of each channel, hue/saturation/value, and Exposure Value which is measured in stops relative to 18% grey.
```
Addressed in [2f8ad2f](https://github.com/GafferHQ/gaffer/pull/6483/commits/2f8ad2fd111a17cd28e3938fe88a31f6fe9ff0f1)

danieldresser-ie commented on src/GafferImageUI/ColorInspectorTool.cpp:
```diff
@@ -602,10 +602,7 @@ class V2iGadget : public GafferUI::Gadget
-					renderLine2D( style, point - V2f( 0, threshold.y ), point - V2f( 0, 2.5 * screenScale.y ), screenScale.x * 2.0f, black );
-					renderLine2D( style, point + V2f( 0, threshold.y ), point + V2f( 0, 2.5 * screenScale.y ), screenScale.x * 2.0f, black );
+					renderCircle2D( style, point, 3.5f * screenScale, screenScale.x * 2.0f, black );
 					renderCircle2D( style, point, 2.5f * screenScale, screenScale.x * 2.0f, Color4f( 0.8, 0.8, 0.8, 1.0 ) );
```
I've made it white, haven't yet increased size, see how you feel. I haven't made the rectangle inspector white ... do you want me to do that for consistency? With it being larger, it doesn't seem to need to be that bright to stand out?

danieldresser-ie commented on python/GafferImageUI/ColorInspectorToolUI.py:
```diff
@@ -587,6 +601,9 @@ def __mouseMove( self, viewportGadget, event ) :
 
 	def __buttonPress( self, viewportGadget, event ) :
+		if not self.getPlug().node()["active"].getValue():
+			return False
```
While reworking this, I ended up going ahead and doing the refactor to move the signal handling code from _ColorInspectorPlugValueWidget to _ColorInspectorsPlugValueWidget. I do feel that having the signal handling on the single parent plug makes more sense than having it on every child plug, but only enabled if the name is equal to the default name. But I hope I haven't created confusion by doing this now: [08389d1](https://github.com/GafferHQ/gaffer/pull/6483/commits/08389d1087901a8e72d48cacdd2049e808c8ea50)

danieldresser-ie commented:
I've centered the color inspector status bar - double check it looks right to you, but seems fine to me, and allows just removing some weird alignment stuff.

The rect gadgets didn't look much different to me whether they were white or 80% grey, so I've gone with white for consistency. I'm now using the most recent version of Murray's icon ( aside from reverting a change to the "foreground" global swatch which I assume was inadvertent in all the passing back and forth ).

I've squashed all changes down - everything was quite interleaved between what was "just get it working after moving the code" and what was a hypothetical independent improvement ... I think if I needed to produce a super-clean history, squashing might have been the first step before trying to split things apart ... but hopefully that won't be necessary.
